presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java

```diff
@@ -1713,10 +1714,58 @@ private PhysicalOperation planGlobalAggregation(int operatorId, AggregationNode
         }
 
-        private PhysicalOperation planGroupByAggregation(AggregationNode node, PhysicalOperation source, int operatorId, Optional<Integer> defaultMaskChannel)
+        private PhysicalOperation planGroupingSetsAggregation(AggregationNode node, PhysicalOperation source, LocalExecutionPlanContext context, Optional<Integer> defaultMaskChannel)
         {
-            List<Symbol> groupBySymbols = node.getGroupBy();
+            List<List<Symbol>> groupingSets = node.getGroupingSets();
 
+            // groupIdOperator is not required for finals or queries with a single grouping set
+            if (groupingSets.size() == 1 || node.getStep() == Step.FINAL) {
+                List<Integer> groupByChannels = getChannelsForSymbols(node.getAllGroupingColumns(), source.getLayout());
+                return planGroupByAggregation(node.getAllGroupingColumns(), groupByChannels, node.getHashSymbol().isPresent(), node, source, context.getNextOperatorId(), defaultMaskChannel);
+            }
+
+            List<List<Integer>> groupingSetChannels = groupingSets.stream()
+                    .map(groupingSet -> getChannelsForSymbols(groupingSet, source.getLayout()))
+                    .collect(toImmutableList());
+            OperatorFactory groupIdOperatorFactory = new GroupIdOperator.GroupIdOperatorFactory(context.getNextOperatorId(), source.types, groupingSetChannels);
+
+            // create layout with groupId as the first channel
+            checkState(node.getGroupIdSymbol().isPresent());
+            Symbol groupIdSymbol = node.getGroupIdSymbol().get();
+
+            ImmutableMap.Builder<Symbol, Integer> newLayoutBuilder = ImmutableMap.builder();
+            newLayoutBuilder.put(groupIdSymbol, 0);
+            for (Map.Entry<Symbol, Integer> symbolChannelEntry : source.getLayout().entrySet()) {
+                newLayoutBuilder.put(symbolChannelEntry.getKey(), symbolChannelEntry.getValue() + 1);
+            }
+            Map<Symbol, Integer> newLayout = newLayoutBuilder.build();
+
+            PhysicalOperation groupIdSource = new PhysicalOperation(groupIdOperatorFactory, newLayout, source);
+
+            // flatten grouping sets into a list of group by columns, add groupId to the groupByColumns
+            ImmutableList<Symbol> groupBySymbols = ImmutableList.<Symbol>builder()
+                    .add(groupIdSymbol)
+                    .addAll(groupingSets.stream()
+                            .flatMap(Collection::stream)
+                            .distinct()
+                            .iterator())
+                    .build();
+
+            List<Integer> groupByChannels = getChannelsForSymbols(groupBySymbols, newLayout);
+
+            // hashes are not precomputed in the case of multiple grouping sets
+            return planGroupByAggregation(groupBySymbols, groupByChannels, false, node, groupIdSource, context.getNextOperatorId(), defaultMaskChannel);
```

Comment:
I would create a new `node` and replace the properties as needed. Having a function that takes `node`, plus some argument that overrides some of the `node`'s field seems hacky.